_published/0013-authentication.md

```diff
@@ -0,0 +1,177 @@
+    levels
+
+- Piggybacking on existing password and user recovery routines results in a
+  DRY style system architecture
+
+- Near perfectly mirrored compliance with top level organisational
+  security levels
+
+## Explanation
+
+By using the OIDC standard an authentication & authorisation system will be
+built in a minimalist fashion that offloads most of the work involved in
+administration to other pre-existing platforms.
+
+![Connectivity visualised](../assets/auth-map-v2.PNG)
+
+### The Access Token
+
+A OIDC type access token can be used for authentication. The custom claims
+(requires an IDP with support for RBAC) can be used to specify access levels.
+To keep the scope manageable the task of dividing up and refining access
+levels and permission will be left to a later date. For this RFC all kinds
+of read permissions won't require valid tokens, but all write type requests
+will require valid ID-tokens that have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[Access Token](https://auth0.com/docs/security/tokens/access-tokens).
+The access token is signed by the source issuer. This signature is then used to
+validate the token. See
+[validating an OIDC Access Token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens).
+
+### Session Management
+
+The wharf-api is currently stateless and does not use sessions. To keep this
+design tokens will be sent in every request to the server.
+
+The option of introducing sessions as a way to store tokens may be reconsidered
+at a later point. There seems to be a potential performance gain.
+
+### Local development vs production environments
+
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+For use in offline developer environments it should be possible to switch on a
+developer mode switch that allows bypassing validation of access
+tokens such that a valid login is not required.
+
+
+### SSL Config Required as a part of OIDC
+
+For OIDC ssl is required. SSL certs will be inserted into the traefik config
+and generated with openSSL as part of the repo setup.
+
+### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. See
+[validating an access token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens)
+
+## Compatibility
+
+Expect no compatibility between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
```

Comment:
I agree with this. I think there will/should be an on/off switch for OIDC. If OIDC is off then basic auth could be enabled or disabled separately. This could cause some strange issues if both are enabled but dosent seem to be a real relevant problem that needs asap handling.